### PR TITLE
update to gym v0.26 Render API

### PR DESCRIPTION
**Pull request by tlpss**
Updates the `vecenv` classes to the new gym render API (no mode argument in `render` function, render mode is specified during init). 
To keep the `VecEnv`API backwards compatible, the mode argument is still accepted but it has to be the same as the environment `render_mode` attribute.

### CODE REVIEW

File: stable_baselines3/common/env_checker.py [hunk 1]
```diff
@@ -285,34 +285,22 @@ def _check_spaces(env: gym.Env) -> None:
...
-        if headless and "human" in render_modes:
-            render_modes.remove("human")
-        # Check all declared render modes
-        for render_mode in render_modes:
-            env.render(mode=render_mode)
-        env.close()
+    # render_modes = env.metadata.get("render.modes")
+    # TODO: if we want to check all render modes,
+    # we need to initialize new environments so the class should be passed as argument.
+    if env.render_mode:
```
Comment (araffin) [inline stable_baselines3/common/env_checker.py:328]: why did you remove the available render mode declaration?
they are not longer declared in gym 0.26?

File: stable_baselines3/common/vec_env/base_vec_env.py [hunk 2]
```diff
@@ -55,10 +54,17 @@ class VecEnv(ABC):
...
+        self,
+        num_envs: int,
+        observation_space: gym.spaces.Space,
+        action_space: gym.spaces.Space,
+        render_mode: Optional[str] = None,
+    ):
         self.num_envs = num_envs
         self.observation_space = observation_space
         self.action_space = action_space
+        self.render_mode = render_mode
```
Comment (araffin) [inline stable_baselines3/common/vec_env/base_vec_env.py:68]: as quickly discussed in https://github.com/DLR-RM/stable-baselines3/pull/780#issuecomment-1302090826, we could do `self.render_mode = render_mode or "rbg_array"`, no?
and keep the rest of the code?

File: stable_baselines3/common/vec_env/dummy_vec_env.py [hunk 3]
```diff
@@ -78,10 +78,10 @@ def close(self) -> None:
         for env in self.envs:
             env.close()
 
-    def get_images(self) -> Sequence[np.ndarray]:
-        return [env.render(mode="rgb_array") for env in self.envs]
+    def get_render_output(self) -> Sequence[Optional[np.ndarray]]:
```
Comment (araffin) [inline stable_baselines3/common/vec_env/dummy_vec_env.py]: I think we should keep `get_images` but raise an error if `render_mode != "rgb_array"`, until gym allows multiple render modes again, see https://github.com/Farama-Foundation/Gymnasium/issues/100

File: stable_baselines3/common/vec_env/subproc_vec_env.py [hunk 4]
```diff
@@ -155,13 +160,12 @@ def close(self) -> None:
             process.join()
         self.closed = True
 
-    def get_images(self) -> Sequence[np.ndarray]:
+    def get_render_output(self) -> Sequence[Optional[np.ndarray]]:
```
Comment (araffin) [inline stable_baselines3/common/vec_env/subproc_vec_env.py]: same, we should keep `get_images` but raise an error if `render_mode` is not `"rgb_array"`

File: stable_baselines3/common/vec_env/base_vec_env.py (hunk 2 above)
Comment (araffin) [reply stable_baselines3/common/vec_env/base_vec_env.py:68]: actually, this should be done in `make_vec_env` and when creating the env from a string

File: stable_baselines3/common/env_checker.py (hunk 1 above)
Comment (tlpss) [reply stable_baselines3/common/env_checker.py:328]: They are still there ( I think), but unless we can instantiate new class instances we cannot test any other mode than the one from the current instance.

File: stable_baselines3/common/vec_env/base_vec_env.py (hunk 2 above)
Comment (tlpss) [reply stable_baselines3/common/vec_env/base_vec_env.py:68]: > actually, this should be done in `make_vec_env` and when creating the env from a string


Yes we can do that when creating from string, but once a gym class instance is passed we can only 'take note' of the render mode I think?

File: stable_baselines3/common/vec_env/dummy_vec_env.py (hunk 3 above)
Comment (tlpss) [reply stable_baselines3/common/vec_env/dummy_vec_env.py]: Works for me!

File: stable_baselines3/common/vec_env/base_vec_env.py (hunk 2 above)
Comment (araffin) [reply stable_baselines3/common/vec_env/base_vec_env.py:68]: >  we can only 'take note' of the render mode I think?

yes and we do the assert only when render is called

File: stable_baselines3/common/env_checker.py (hunk 1 above)
Comment (araffin) [reply stable_baselines3/common/env_checker.py:328]: > ny other mode than the one from the current instance.

yes, I'm more talking about what we use to test with the warning (check if any render mode is declared and output a warning otherwise)

Comment (tlpss) [reply stable_baselines3/common/env_checker.py:328]: ah yes, my bad. I was too hasty apparently, will bring that check back

File: stable_baselines3/common/vec_env/dummy_vec_env.py (hunk 3 above)
Comment (tlpss) [reply stable_baselines3/common/vec_env/dummy_vec_env.py]: @araffin , how do you want the render method of the VecEnv to behave then? Should it raise an Error if the render_mode is not rgb_array (in the future if not one of the modes is rgb_array)? Or do you want to the Error in the get_images method?

File: stable_baselines3/common/env_util.py [hunk 5]
```diff
@@ -74,14 +74,18 @@ def make_vec_env(
...
+
     vec_env_kwargs = {} if vec_env_kwargs is None else vec_env_kwargs
     monitor_kwargs = {} if monitor_kwargs is None else monitor_kwargs
     wrapper_kwargs = {} if wrapper_kwargs is None else wrapper_kwargs
 
     def make_env(rank):
         def _init():
             if isinstance(env_id, str):
-                env = gym.make(env_id, **env_kwargs)
+                # if the render mode was not specified, we set it to `rgb_array` as default.
```
Comment (araffin) [inline stable_baselines3/common/env_util.py:90]: I need to check, but is `render_mode` argument part of the gym api?

If not, then we should do that unless we know this argument exists.

Comment (tlpss) [reply stable_baselines3/common/env_util.py:90]: @araffin you're right, it is not required for each class to have this argument in the init function (it is a guaranteed to exist though, see https://github.com/Farama-Foundation/Gymnasium/blob/f037fdea5a0adca5573aa5509bdc77bf48603416/gymnasium/core.py#L51) 
But their migration says this is the way to set the render mode. 

We could wrap it with a try except block to accomodate for environments that don't allow to specify the render mode?

File: stable_baselines3/common/vec_env/dummy_vec_env.py (hunk 3 above)
Comment (tlpss) [reply stable_baselines3/common/vec_env/dummy_vec_env.py]: @araffin, gentle ping

File: stable_baselines3/common/env_util.py (hunk 5 above)
Comment (araffin) [reply stable_baselines3/common/env_util.py:90]: > it is a guaranteed to exist though

then should be fine (we might revert it back later in case it causes too much troubles)

File: stable_baselines3/common/vec_env/dummy_vec_env.py (hunk 3 above)
Comment (araffin) [reply stable_baselines3/common/vec_env/dummy_vec_env.py]: >  Should it raise an Error if the render_mode is not rgb_array (in the future if not one of the modes is rgb_array)? 

probably a warning would be better? (so a warning and do nothing)
crashing someone's code because one tries to render doesn't look like a user-friendly behavior, no?

Comment (tlpss) [reply stable_baselines3/common/vec_env/dummy_vec_env.py]: > I think we should keep get_images but raise an error if render_mode != "rgb_array"

Yes well I was trying to infer what you ment by this. 

I added the warning and reverted the name change. Let me know if there is anything else you would like to see different

File: stable_baselines3/common/vec_env/base_vec_env.py [hunk 6]
```diff
@@ -162,35 +168,44 @@ def step(self, actions: np.ndarray) -> VecEnvStepReturn:
         self.step_async(actions)
         return self.step_wait()
 
-    def get_images(self) -> Sequence[np.ndarray]:
+    def get_images(self) -> Sequence[Optional[np.ndarray]]:
         """
-        Return RGB images from each environment
+        Return Render output from each environment
```
Comment (araffin) [inline stable_baselines3/common/vec_env/base_vec_env.py]:
```suggestion
        Return RGB images from each environment when available
```

File: stable_baselines3/common/vec_env/base_vec_env.py [hunk 7]
```diff
@@ -162,35 +168,44 @@ def step(self, actions: np.ndarray) -> VecEnvStepReturn:
...
-        except NotImplementedError:
-            warnings.warn(f"Render not defined for {self}")
-            return
 
-        # Create a big image by tiling images from subprocesses
-        bigimg = tile_images(imgs)
-        if mode == "human":
-            import cv2  # pytype:disable=import-error
+        if mode and self.render_mode != mode:
+            raise ValueError(
```
Comment (araffin) [inline stable_baselines3/common/vec_env/base_vec_env.py]: should be a warning I think (and return right after).
That way it does not crash.

Comment (araffin) [reply stable_baselines3/common/vec_env/base_vec_env.py]: btw, I think we should support `mode="human"` when `render_mode="rgb_array"` because it is how it works with subprocess env anyway.

File: stable_baselines3/common/vec_env/base_vec_env.py [hunk 8]
```diff
@@ -162,35 +168,44 @@ def step(self, actions: np.ndarray) -> VecEnvStepReturn:
...
+            bigimg = tile_images(images)
             return bigimg
+
+        elif mode == "rgb_array_list":
+            # TODO: a new 'rgb_array_list' mode has been defined and should be handled.
+            raise NotImplementedError("This mode has not yet been implemented in Stable Baselines.")
+
         else:
-            raise NotImplementedError(f"Render mode {mode} is not supported by VecEnvs")
+            # other render methods are simply ignored.
```
Comment (araffin) [inline stable_baselines3/common/vec_env/base_vec_env.py]: actually no, we should display the big image I think (if available, i.e. render_mode=`rgb_array`).
I know this might seems contradictory but this has to do with the fact that we are using multiple envs, not only one (and the limitation comes from https://github.com/Farama-Foundation/Gymnasium/issues/100)

File: stable_baselines3/common/vec_env/base_vec_env.py [hunk 9]
```diff
@@ -162,35 +168,44 @@ def step(self, actions: np.ndarray) -> VecEnvStepReturn:
...
+            raise ValueError(
+                f"""starting from gym v0.26, render modes are determined during the initialization of the environment.
+                We allow to pass a mode argument to maintain a backwards compatible VecEnv API, but the mode ({mode}) 
+                has to be the same as the environment render mode ({self.render_mode}) whichs is not the case."""
+            )
 
-            cv2.imshow("vecenv", bigimg[:, :, ::-1])
-            cv2.waitKey(1)
-        elif mode == "rgb_array":
+        mode = self.render_mode
```
Comment (araffin) [inline stable_baselines3/common/vec_env/base_vec_env.py:191]: hmm, should be `self.get_attr("render_mode")[0]` I guess.
you can either remove `self.render_mode` or better set it to `self.get_attr("render_mode")[0]` when initializing the envs (and so remove it anyway from the VecEnv constructor).

Currently, having `render_mode`  in VecEnv constructor, is kind of a duplicate, no?

File: stable_baselines3/common/vec_env/subproc_vec_env.py [hunk 10]
```diff
@@ -45,13 +45,15 @@ def _worker(  # noqa: C901
...
             elif cmd == "render":
-                remote.send(env.render(data))
+                remote.send(env.render())
             elif cmd == "close":
                 env.close()
                 remote.close()
                 break
             elif cmd == "get_spaces":
                 remote.send((env.observation_space, env.action_space))
+            elif cmd == "get_render_mode":
```
Comment (araffin) [inline stable_baselines3/common/vec_env/subproc_vec_env.py]: you can probably re-use `cmd == "get_attr"`, no?

File: tests/test_vec_check_nan.py [hunk 11]
```diff
@@ -30,7 +30,7 @@ def step(action):
     def reset():
         return [0.0], {}
 
-    def render(self, mode="human", close=False):
+    def render(self, close=False):
```
Comment (araffin) [inline tests/test_vec_check_nan.py]: we can even remove `close=False` I think

File: stable_baselines3/common/vec_env/dummy_vec_env.py [hunk 12]
```diff
@@ -89,10 +89,14 @@ def close(self) -> None:
         for env in self.envs:
             env.close()
 
-    def get_images(self) -> Sequence[np.ndarray]:
-        return [env.render(mode="rgb_array") for env in self.envs]
+    def get_images(self) -> Sequence[Optional[np.ndarray]]:
+        if self.envs[0].render_mode != "rgb_array":
```
Comment (araffin) [inline stable_baselines3/common/vec_env/dummy_vec_env.py]: why not re-use `self.render_mode` ?

File: stable_baselines3/common/vec_env/subproc_vec_env.py [hunk 13]
```diff
@@ -155,13 +160,12 @@ def close(self) -> None:
...
 
-    def get_images(self) -> Sequence[np.ndarray]:
+    def get_images(self) -> Sequence[Optional[np.ndarray]]:
         for pipe in self.remotes:
-            # gather images from subprocesses
-            # `mode` will be taken into account later
-            pipe.send(("render", "rgb_array"))
-        imgs = [pipe.recv() for pipe in self.remotes]
-        return imgs
+            # gather render return from subprocesses
```
Comment (araffin) [inline stable_baselines3/common/vec_env/subproc_vec_env.py:169]: I guess you can do the same check/warning as for the `DummyVecEnv` here

File: stable_baselines3/common/vec_env/base_vec_env.py (hunk 8 above)
Comment (tlpss) [reply stable_baselines3/common/vec_env/base_vec_env.py]: So if I have understood you correctly, we  should return the tiled image for both 'human' and 'rgb_array'? If not let me know and I can change that again.

Comment (araffin) [reply stable_baselines3/common/vec_env/base_vec_env.py]: I did the change ;) (so show the image when possible)

**Comment by araffin** [comment]
I think we are almost good to go, what is missing is to test and debug with different envs/situation.

There is one case that I'm not will work, when `env.render_mode == "human"`, it might crash anyway when using subprocesses, but it might work too. (need to be checked)

EDIT: the documentation need to be updated too to explain the behaviors, and recommend `rgb_array` as a good default ;)

**Comment by tlpss** [comment]
> I think we are almost good to go, what is missing is to test and debug with different envs/situation.

could you briefly explain what you expect from this? The render API of the vecenv has not changed I think? so do you want to add new tests or should I modify existing tests?

**Comment by araffin** [comment]
Hello,
i mainly meant checking manually with different envs and scenarii, we cannot really check that on the ci server.

**Comment by tlpss** [comment]
> Hello, i mainly meant checking manually with different envs and scenarii, we cannot really check that on the ci server.

Okay, I can test for a few enviroments manually to see if the output is what I expect

Could you also point me to the place(s) where I need to change the docs?

**Comment by araffin** [comment]
> Could you also point me to the place(s) where I need to change the docs?

I'm not sure yet. Probably the vec env page to explain difference in behavior vs gym api, and then the example page where we do some recording/render I think.

**Comment by araffin** [comment]
So, I gave it a try and fixed some bugs along the way.
I also found one bug in gym due to shared memory, those two snippets creates only one window (can be solved using subprocess):

cc @pseudo-rnd-thoughts

```python
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.vec_env import SubprocVecEnv


if __name__ == "__main__":
    n_envs = 2
    env = make_vec_env(
        "CartPole-v1",
        n_envs,
        # vec_env_cls=SubprocVecEnv,
        env_kwargs=dict(render_mode="human")
    )

    env.reset()
    env.render()

    while True:
        env.step([env.action_space.sample() for _ in range(n_envs)])
        env.render()

```


```python
import gym

env1 = gym.make("CartPole-v1", render_mode="human")
env2 = gym.make("CartPole-v1", render_mode="human")

envs = [env1, env2]

for env in envs:
    env.reset()


while True:
    for env in envs:
        env.step(env.action_space.sample())
        env.render()

```

**Comment by tlpss** [comment]
great @araffin, I'll have a look at the docs.

**Comment by tlpss** [comment]
> great @araffin, I'll have a look at the docs.

@araffin 
I think we might want to introduce a new page in the docs to explain the differences between the VecEnv interface and the new gym v0.26 interface? 
And then we probably need to update the examples section and add an explicit example to show the difference between interacting with a gym>0.26 interface and a vecenv interface?

Do you think of other places that should be changed? 
and are there any tests for the code snippets in the docs or do we have to test them manually?

**Comment by araffin** [comment]
> I think we might want to introduce a new page in the docs to explain the differences between the VecEnv interface and the new gym v0.26 interface?

why not a section in the vec env page?
it's gonna be too big?

>  and are there any tests for the code snippets in the docs or do we have to test them manually?

we need to test them manually (too long to run at every push, see https://github.com/DLR-RM/stable-baselines3/issues/14)

**Comment by araffin** [comment]
I'm not sure Github notification worked as I edited the original post.
@pseudo-rnd-thoughts : https://github.com/carlosluis/stable-baselines3/pull/3#issuecomment-1356863808

if you have already seen it, then please ignore my message.

**Comment by araffin** [comment]
@tlpss I added the doc and will merge this branch (I tested it in different scenarios and also added tests that are run only when display is available).
In the meantime, could you read what I wrote and check that nothing is missing/things are clear?
(you can do a PR or use this one if you have suggestions/remarks)

**Comment by tlpss** [comment]
> @tlpss I added the doc and will merge this branch (I tested it in different scenarios and also added tests that are run only when display is available)

great!

> could you read what I wrote and check that nothing is missing/things are clear?

I think all differences between the gym API and VecEnv API are clearly listed. 

Only suggestion : I would  add a reference to this section in the quickstart section of the docs?

**Comment by araffin** [comment]
> Only suggestion : I would add a reference to this section in the quickstart section of the docs?

Good point, will do that =)

**Comment by pseudo-rnd-thoughts** [comment]
> So, I gave it a try and fixed some bugs along the way. I also found one bug in gym due to shared memory, those two snippets creates only one window (can be solved using subprocess):
> 
> cc @pseudo-rnd-thoughts
> 
> ```python
> from stable_baselines3.common.env_util import make_vec_env
> from stable_baselines3.common.vec_env import SubprocVecEnv
> 
> 
> if __name__ == "__main__":
>     n_envs = 2
>     env = make_vec_env(
>         "CartPole-v1",
>         n_envs,
>         # vec_env_cls=SubprocVecEnv,
>         env_kwargs=dict(render_mode="human")
>     )
> 
>     env.reset()
>     env.render()
> 
>     while True:
>         env.step([env.action_space.sample() for _ in range(n_envs)])
>         env.render()
> ```
> 
> ```python
> import gym
> 
> env1 = gym.make("CartPole-v1", render_mode="human")
> env2 = gym.make("CartPole-v1", render_mode="human")
> 
> envs = [env1, env2]
> 
> for env in envs:
>     env.reset()
> 
> 
> while True:
>     for env in envs:
>         env.step(env.action_space.sample())
>         env.render()
> ```

Sorry, I must have missed this. 

I will create an issue for it, I wonder if it works with vectorised gym environments
